qutip/core/data/convert.pyx

```diff
@@ -294,28 +318,98 @@ cdef class _to:
+            Each element can individually be a 2- or 3-tuple; they do not need
+            to be all one or the other.
 
+            Elements
+            ........
+            condition : callable (object) -> Data
+                function determining if that object can be made to a data layer
+                using this creator.
+
+            creator function: callable (object, shape) -> Data
+                The creator function.  This should take an object and a shape
+                and output a data-layer object. The object can be anything that
+                The condition function returned `True` when tested.
+
+            priority : positive real, optional (1)
+                The priority associated with this creation. Higher priority
+                conditions will be tested first and the first valid creator
+                (condition(object) == True) will handle the creation.
+                Priority 100 is used for already data-layer entry.
+                80 for object that have a direct data-layer equivalent
+                (scipy.sparse.csr_matrix, numpy.ndarray)
+                0 is the numpy fallback, making a data-layer of anything that
+                numpy can make to complex array.
+        """
+        _creators = []
+        for creator in creators:
+            if len(creator) == 2:
+                condition, create = creator
+                priority = 10
+            elif len(creator) == 3:
+                condition, create, priority = creator
+            if not callable(condition):
+                raise TypeError(str(condition) + " is not a callable")
+            if not callable(create):
+                raise TypeError(str(create) + " is not a callable")
+            _creators.append((condition, create, priority))
+        self._creators = sorted(_creators,
+                                key=lambda creator: creator[2],
+                                reverse=True)
+
+    def __call__(self, arg, shape=None):
+        for condition, create, _ in self._creators:
+            if condition(arg):
+                return create(arg, shape)
 
 to = _to()
 create = _create()
+
+import qutip.core.data as _data
+import numpy as np
+import scipy.sparse
+
+create.add_creators([
+    (
+        lambda arg: isinstance(arg, _data.base.Data),
+        lambda arg, shape: arg.copy(),
+        100
+    ),
+    (
+        lambda arg: scipy.sparse.isspmatrix_csr(arg),
+        lambda arg, shape: _data.CSR(arg, shape=shape),
+        80
+    ),
+    (
+        lambda arg: isinstance(arg, np.ndarray),
+        lambda arg, shape: _data.Dense(arg, shape=shape, copy=False),
+        80
+    ),
+    (
+        lambda arg: scipy.sparse.issparse(arg),
+        lambda arg, shape: _data.CSR(arg.tocsr(), shape=shape),
+        10
+    ),
+    (
+        lambda arg: True,
+        lambda arg, shape: _data.Dense(np.array(arg, dtype=np.complex128),
+                                       shape=shape, copy=False),
```

Comment:
This looks like it's asking for trouble.  Might be better to handle the "fall-through" case explicitly in `create`.